File: 34. Find First and Last Position of Element in Sorted Array/34. Find First and Last Position of Element in Sorted Array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = findBound(nums, target, true);
        if (left == -1) return {-1, -1};
        int right = findBound(nums, target, false);
        return {left, right};
    }
    
private:
    int findBound(vector<int>& nums, int target, bool isLeft) {
        int left = 0, right = nums.size() - 1;
        int result = -1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                result = mid;
                if (isLeft) right = mid - 1;
                else left = mid + 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return result;
    }
};

File: 34. Find First and Last Position of Element in Sorted Array/34. Find First and Last Position of Element in Sorted Array.rs
impl Solution {
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let left = Self::find_bound(&nums, target, true);
        if left == -1 {
            return vec![-1, -1];
        }
        let right = Self::find_bound(&nums, target, false);
        vec![left, right]
    }
    
    fn find_bound(nums: &Vec<i32>, target: i32, is_left: bool) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        let mut result = -1;
        
        while left <= right {
            let mid = left + (right - left) / 2;
            if nums[mid as usize] == target {
                result = mid;
                if is_left {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else if nums[mid as usize] < target {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        result
    }
}